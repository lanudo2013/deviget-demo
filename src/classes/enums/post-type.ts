export enum PostType {
    IMAGE, VIDEO, LINK, SELF
}

export function fromPostType(val: string) : PostType | null {
    if (val) {
        val = val.toLowerCase();
        switch(val) {
            case 'image':
                return PostType.IMAGE;
            case 'hosted:video':
                return PostType.VIDEO;
            case 'self':
                return PostType.SELF;
            case 'link':
                return PostType.LINK;
            default:
                return null;
        }
    }
    return PostType.IMAGE;
}