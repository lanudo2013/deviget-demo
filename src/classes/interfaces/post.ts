export interface Post {
    title: string;
    thumbnailUrl: string;
    createdTime: Date;
    author: string;
    id: string;
    fullPictureUrl: string;
    subreddit: string;
    thumbnailDims?: {width: number, height: number};
    numberOfComments: number;
}