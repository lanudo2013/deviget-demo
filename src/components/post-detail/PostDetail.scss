@import '../../variables';
.PostDetailContainer {
    display: flex;
    flex-direction: column;
    flex: 1 1;
    padding-left: 2em;
    padding-top: 5px;   
    padding-right: 1em;
    overflow-y: auto;
    height: 100%;
    position: relative;
    @media screen and (max-width: $MOBILE_WIDTH_LIMIT) {
        padding-left: 0em;
    }
    .AuthorContainer {
        display: block;
        margin-bottom: 0.8em;
        height: 2em;
        .Author {
            text-align: left;
            font-size: 1.2em;
            display: inline-block;
            padding-left: 3em;
            border-radius: 12px;
            font-weight: bold;
            background: #e6d9d9;
            padding-right: 3em;
            color: #797979;
            align-items: center;
            padding-top: 5px;
            padding-bottom: 5px;
        }
    }
    
    .InnerContainer {
        padding-left: 8%;
        padding-right: 8%;
        //min-height: calc(100% - 6em);
        margin-bottom: 1.6em;
        flex: 1;
        padding-top: 1em;
        box-shadow: 2px 2px 12px -6px #000000;
        border: solid 0.1px #ded8d8;
        flex-direction: column;
        display: flex;
    }
    .Title {
        text-align: center;
        font-size: 1.6em;
        color: grey;
        border-bottom: solid 0.3px #cabebe;
        font-weight: bold;
        margin-bottom: 15px;
        padding-left: 8px;
        padding-right: 8px;
        padding-bottom: 12px;
    }
    .Body {
        width: 100%;
        display: flex;
        flex: 1;
        flex-direction: column;
        margin-bottom: 2em;
        justify-content: center;
        align-items: center;
        .Body-content {
            flex: 1;
            width: 100%;
            display: flex;
            justify-content: center;
            position: relative;
            flex-direction: column;
            &.NotCentered {
                justify-content: flex-start;
            }
        }
        .PostPicture {
            background-repeat: no-repeat;
            background-position: center;
            background-size: contain;
            min-height: 300px;
            min-width: 300px;
            flex: 0.9;
        }
        
        .Video {
            position: absolute;
            height: 100%;
            max-width: 100%;
            width: auto;
            transform: translate(-50%, 0);
            left: 50%;
        }
        .EmbedVideo {
            align-items: center;
            justify-content: flex-start;
        }
        .Url {
            text-align: center;
            font-size: 1.5em;
        }
    }
}