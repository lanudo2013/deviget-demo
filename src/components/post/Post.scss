.PostContainer {
    display: flex;
    flex-direction: column;
    margin-left: 10px;
    border-radius: 20px;
    margin-right: 10px;
    margin-top: 3px;
    padding: 20px;
    margin-bottom: 10px;
    box-shadow: 2px 1px 3px 2px rgb(191 191 191 / 42%);
    animation-duration: 0.8s !important;

    .Header {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        .Author {
            font-size: 20px;
            font-weight: bold;
            line-height: 10px;
        }
    }
    .Body {
        min-height: 90px;
        display: flex;
        margin-top: 5px;
        align-items: center;
        border: solid 0px black;
        .Thumbnail {
        }
        .Title {
            vertical-align: center;
            margin-left: 10px;
            align-self: center;
        }
    }
    .Footer {
        margin-top: 7px;
        display: flex;
        justify-content: space-between;
        .DismissButton {
            padding-top: 1px;
            padding-bottom: 1px;
        }
        .CommentsNum {
            color: rgb(165, 165, 41);
            font-size: 16px;
            font-weight: bold;
        }
    }
}