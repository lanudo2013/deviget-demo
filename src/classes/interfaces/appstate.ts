import { DismissData } from "./dismiss-data";
import { Post } from "./post";

export interface AppState {
    posts: Post[];
    selectedPost: Post | undefined | null;
    fetchingPosts: boolean;
    currentError: string;
    dismissData: DismissData | undefined;
    postsRead: string[];
}