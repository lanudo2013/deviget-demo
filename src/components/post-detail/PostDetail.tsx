import React from 'react';
import { useDispatch, useSelector } from "react-redux";
import { PostType } from '../../classes/enums/post-type';
import { AppState } from "../../classes/interfaces/appstate";
import { PostVideo } from '../../classes/interfaces/post';
import './PostDetail.scss';

interface PostDetailProps {
}
export const PostDetail = function(props: PostDetailProps) {
    const dispatch = useDispatch();
    const post = useSelector((state: AppState) => state.selectedPost);

    const htmlDecode = React.useCallback((input: string) => {
        const e = document.createElement('div');
        e.innerHTML = input;
        return e.childNodes.length === 0 ? "" : (e.childNodes[0].nodeValue || '');
    }, []);
    
    const getPostBody = React.useCallback(() => {
        if (!post) {
            return null;
        }
        const vData = post.videoData || {} as PostVideo;
        switch(post.postType) {
            case PostType.IMAGE:
                return <div className="Img-container"><div className="PostPicture" style={{backgroundImage: `url(${post.postUrl})`}}/></div>;
            case PostType.VIDEO:
                return <div className="Img-container"><video src={vData.url} className="Video" autoPlay controls/></div>;
            case PostType.LINK:
                return <div className="Img-container"><a href={post.postUrl} className="Url" target="_blank">{post.postUrl}</a></div>;
            case PostType.SELF:
                return <div className="Img-container" dangerouslySetInnerHTML={{__html: htmlDecode(post.postHtml)}}></div>;
            default:
                return null;
        }
    }, [post]);
    if (!post) {
        return null;
    }

    const body = getPostBody();
    return <div className="PostDetailContainer">
        <span className="Author">{post.author}</span>
        <div className="InnerContainer" style={!body ? {justifyContent: 'center'} : {}}>
            <span className="Title" style={!body ? {fontSize: '2.4em'} : {}}>{post.title}</span>
            {
                body ? <div className="Body">{body}</div> : null
            }
        </div>
       
    </div>;
};