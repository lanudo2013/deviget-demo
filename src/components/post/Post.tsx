import { Post } from "../../classes/interfaces/post";
import React, { useState } from 'react';
import moment from  'moment';
import './Post.scss';
import { Button } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/CancelOutlined';

const DAY_TIME = 1000 * 60 * 60 * 24;
const HOUR_TIME = 1000 * 60 * 60;
const MINUTE_TIME = 1000 * 60;

interface PostProps {
    post: Post;
    currentDate: Date;
    onPress: (p: Post) => void;
    onDismiss: (post: Post) => void;
}
export const PostUI = function(props: PostProps) {
    const [pressAnimation, setPressAnimation] = useState<boolean>(false);
    const [dismissAnimation, setDismissAnimation] = useState<boolean>(false);
    const formatCreated = React.useCallback((dt: Date) => {
        const now = props.currentDate;
        const diff = now.getTime() - dt.getTime();
        if (diff < HOUR_TIME) {
            return `${Math.floor(diff / MINUTE_TIME)} minutes ago`;
        } else if (diff < DAY_TIME) {
            return `${Math.floor(diff / HOUR_TIME)} hours ago`;
        } else if (diff < 5 * DAY_TIME) {
            return `${Math.floor(diff / DAY_TIME)} days, ${(diff % DAY_TIME) / HOUR_TIME} hours ago`;
        } else {
            return `${Math.floor(diff / DAY_TIME)} days ago`;
        }
    }, []);

    const dismissButtonStyle = React.useMemo(() => ({paddingTop: '1px', paddingBottom: '1px', fontSize: '12px'}), []);

    const p = props.post;
    const pressPost = React.useCallback((e) => {
        setPressAnimation(true);
        props.onPress && props.onPress(p);
    }, [props.onPress, p]);

    
    const animEnds = React.useCallback((e) => {
        setPressAnimation(false);
        if (dismissAnimation) {
            props.onDismiss && props.onDismiss(p);
        }
    }, [p, props.onDismiss]);

    const pressDismiss = React.useCallback((e: any) => {
        e.stopPropagation();
        setDismissAnimation(true);
    }, []);

    if (!props.post) {
        return null;
    }
    


    return <div className={'PostContainer ' + (pressAnimation ? ' animate__animated animate__headShake' : '') + (dismissAnimation ? 'animate__animated animate__fadeOut' : '')} onClick={pressPost} onAnimationEnd={animEnds}>
        <div className="Header">
            <span className="Author">{p.author}</span>
            <span className="CreatedAt">{formatCreated(p.createdTime)}</span>
        </div>
        <div className="Body">
            {p.thumbnailUrl ? <img src={p.thumbnailUrl} width={100} height={100} className="Thumbnail" /> : null}
            <span className="Title">{p.title}</span>
        </div>
        <div className="Footer">
            <Button onClick={pressDismiss} color="default" size={'small'} style={dismissButtonStyle} variant="contained" startIcon={<DeleteIcon />}>Dismiss post</Button>
            <span className="CommentsNum">{p.numberOfComments} comment/s</span>
        </div>
    </div>;
};