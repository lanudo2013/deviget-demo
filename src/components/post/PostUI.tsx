import { Post } from '../../classes/interfaces/post';
import React, { ForwardedRef, RefObject, useEffect, useImperativeHandle, useRef, useState } from 'react';
import moment from 'moment';
import './Post.scss';
import { Button } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/CancelOutlined';
import DeleteOutlineIcon from '@material-ui/icons/DeleteOutline';
import SaveIcon from '@material-ui/icons/SaveOutlined';
import Icon from '@material-ui/core/Icon';
import { Constants } from '../../constants';
import Intl from 'intl';
import { AnyIfEmpty } from 'react-redux';
import { PostType } from '../../classes/enums/post-type';

const DAY_TIME = 1000 * 60 * 60 * 24;
const HOUR_TIME = 1000 * 60 * 60;
const MINUTE_TIME = 1000 * 60;

interface PostProps {
    post: Post;
    currentDate: Date;
    onPress: (p: Post) => void;
    onPressSave?: (p: Post) => void;
    onPressRemoved?: (p: Post) => void;
    read?: boolean;
    canRemove?: boolean;
    saved?: boolean;
    onPressDismiss: (post: Post) => void;
}
export interface PostRef {
    fadeOut: () => Promise<any>;
    id: string;
    slideOut: (delay?: number) => Promise<any>;
}
const PostUIFn = function (props: PostProps, ref1: ForwardedRef<any>) {
    const [pressAnimation, setPressAnimation] = useState<boolean>(false);
    const [fadeOutAnimation, setDismissFadeOutAnimation] = useState<boolean>(false);
    const [slideOutAnimation, setDismissSlideOutAnimation] = useState<boolean>(false);
    const [hidden, setHidden] = useState<boolean>(false);
    const resolveFadeOutAnim = useRef<any>(null);
    const resolveSlideAnim = useRef<any>(null);
    const formatCreatedAt = React.useCallback(
        (dt: Date) => {
            const now = props.currentDate;
            if (now && dt) {
                const diff = now.getTime() + now.getTimezoneOffset() * 1000 * 60 - dt.getTime();
                const days = Math.floor(diff / DAY_TIME);
                const minutes = Math.floor(diff / MINUTE_TIME);
                const hours = Math.floor(diff / HOUR_TIME);
                if (diff > 0) {
                    if (diff < MINUTE_TIME) {
                        return 'Seconds ago';
                    } else if (diff < HOUR_TIME) {
                        return `${minutes} minute${minutes > 1 ? 's' : ''} ago`;
                    } else if (diff < DAY_TIME) {
                        return `${hours} hour${hours > 1 ? 's' : ''} ago`;
                    } else if (diff < 5 * DAY_TIME) {
                        const hoursDiff = Math.floor((diff % DAY_TIME) / HOUR_TIME);
                        if (hoursDiff > 0) {
                            return `${days} day${days > 1 ? 's' : ''}, ${hoursDiff} hour${hoursDiff > 1 ? 's' : ''} ago`;
                        } else {
                            return `${days} day${days > 1 ? 's' : ''} ago`;
                        }
                    } else {
                        return `${days} day${days > 1 ? 's' : ''} ago`;
                    }
                }
            }
            return '';
        },
        [props.currentDate]
    );

    const p = props.post;
    const dismissButtonStyle = React.useMemo(
        () => ({ paddingTop: '1px', paddingBottom: '1px', fontSize: '12px', marginRight: '5px' }),
        []
    );
    useImperativeHandle(
        ref1,
        () => ({
            fadeOut: () => {
                if (resolveFadeOutAnim.current) {
                    return Promise.reject('Animating');
                }

                setDismissFadeOutAnimation(true);
                return new Promise((res) => {
                    resolveFadeOutAnim.current = res;
                });
            },
            slideOut: (delay?: number) => {
                if (resolveSlideAnim.current) {
                    return Promise.reject('Animating');
                }
                if (!delay) {
                    setDismissSlideOutAnimation(true);
                }
                return new Promise((res) => {
                    resolveSlideAnim.current = res;
                    if (delay) {
                        setTimeout(() => {
                            setDismissSlideOutAnimation(true);
                        }, delay);
                    }
                });
            },
            id: p.id
        }),
        [p]
    );

    const formatter = React.useMemo(() => {
        return new Intl.NumberFormat('en-IR');
    }, []);

    const pressPost = React.useCallback(
        (e) => {
            setPressAnimation(true);
            props.onPress && props.onPress(p);
        },
        [props.onPress, p]
    );

    const pressSave = React.useCallback(
        (e: any) => {
            e.stopPropagation();
            props.onPressSave && props.onPressSave(p);
        },
        [props.onPressSave, p]
    );

    const animEnds = React.useCallback(
        (e) => {
            setPressAnimation(false);
            if (fadeOutAnimation && resolveFadeOutAnim.current) {
                resolveFadeOutAnim.current(true);
            } else if (slideOutAnimation && resolveSlideAnim.current) {
                setHidden(true);
                resolveSlideAnim.current(true);
            }
            resolveSlideAnim.current = null;
            resolveFadeOutAnim.current = null;
            setDismissSlideOutAnimation(false);
            setDismissFadeOutAnimation(false);
        },
        [p, fadeOutAnimation, slideOutAnimation]
    );

    const pressDismiss = React.useCallback(
        (e: any) => {
            e.stopPropagation();
            props.onPressDismiss && props.onPressDismiss(p);
        },
        [props.onPressDismiss, p]
    );

    const pressRemoved = React.useCallback(
        (e: any) => {
            e.stopPropagation();
            props.onPressRemoved && props.onPressRemoved(p);
        },
        [props.onPressRemoved, p]
    );

    const animationClassStr = React.useMemo(() => {
        if (pressAnimation) {
            return ' animate__animated animate__headShake';
        } else if (fadeOutAnimation) {
            return ' animate__animated animate__fadeOut';
        } else if (slideOutAnimation) {
            return ' animate__animated animate__slideOutLeft';
        } else {
            return '';
        }
    }, [slideOutAnimation, fadeOutAnimation, pressAnimation]);

    const openFullSizeImage = React.useCallback(
        (e) => {
            const val = p.postUrl || '';
            if (p.postType === PostType.IMAGE && (val.endsWith('.jpg') || val.endsWith('.png') || val.endsWith('.gif') || val.endsWith('.jpeg'))) {
                e.stopPropagation();
                window.open(p.postUrl, '_blank');
            }
        },
        [p]
    );

    if (!props.post) {
        return null;
    }

    return (
        <div
            className={'PostContainer ' + animationClassStr}
            style={hidden ? { visibility: 'hidden' } : {}}
            onClick={pressPost}
            onAnimationEnd={animEnds}
        >
            <div className="Header">
                <div className="Header-left">
                    {!props.read ? <Icon className="ReadIcon">fiber_manual_record</Icon> : null}
                    <span className="Author" style={props.read ? { left: 0 } : null}>
                        {p.author}
                    </span>
                </div>
                <div className="Header-right">
                    {props.saved ? <Icon className="SavedIcon">save</Icon> : null}
                    <span className="CreatedAt">{formatCreatedAt(p.createdTime)}</span>
                </div>
            </div>
            <div className="Body">
                {p.thumbnailUrl ? (
                    <img
                        src={p.thumbnailUrl}
                        width={100}
                        height={100}
                        onClick={openFullSizeImage}
                        className={'Thumbnail ' + (p.postType === PostType.IMAGE ? 'Clickable' : '')}
                    />
                ) : null}
                <span className="Title">{p.title}</span>
            </div>
            <div className="Footer">
                <div className="PostOptions">
                    {!props.saved ? (
                        <>
                            <Button
                                onClick={pressDismiss}
                                color="default"
                                size={'small'}
                                style={dismissButtonStyle}
                                variant="contained"
                                className="DismissButton"
                                startIcon={<DeleteIcon />}
                            >
                                {Constants.APP_MESSAGES.DISMISS_BUTTON}
                            </Button>
                            <Button
                                onClick={pressSave}
                                color="default"
                                size={'small'}
                                style={dismissButtonStyle}
                                variant="contained"
                                className="SaveButton"
                                startIcon={<SaveIcon />}
                            >
                                {Constants.APP_MESSAGES.SAVE_BUTTON}
                            </Button>
                        </>
                    ) : (
                        <>
                            {props.canRemove ? (
                                <Button
                                    onClick={pressRemoved}
                                    color="default"
                                    size={'small'}
                                    style={dismissButtonStyle}
                                    variant="contained"
                                    className="RemoveButton"
                                    startIcon={<DeleteOutlineIcon />}
                                >
                                    {Constants.APP_MESSAGES.REMOVE_BUTTON}
                                </Button>
                            ) : (
                                <Button
                                    onClick={pressDismiss}
                                    color="default"
                                    size={'small'}
                                    style={dismissButtonStyle}
                                    variant="contained"
                                    className="DismissButton"
                                    startIcon={<DeleteIcon />}
                                >
                                    {Constants.APP_MESSAGES.DISMISS_BUTTON}
                                </Button>
                            )}
                        </>
                    )}
                </div>

                <span className="CommentsNum">
                    {p.numberOfComments > 0 ? formatter.format(p.numberOfComments) + ' comment/s' : 'No comments'}
                </span>
            </div>
        </div>
    );
};
export const PostUI = React.forwardRef(PostUIFn);
