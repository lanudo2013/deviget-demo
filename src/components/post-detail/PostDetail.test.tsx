import { PostType } from '../../classes/enums/post-type';
import { Post } from '../../classes/interfaces/post';
import { mount, ReactWrapper, shallow } from "enzyme";
import moment from 'moment';
import React, { useRef } from 'react';
import { useSelector, useDispatch, Provider } from 'react-redux'; 
import { PostDetail } from './PostDetail';
import { AppState } from '../../classes/interfaces/appstate';
import rootReducer from '../../state/rootReducer';
import { createStore } from 'redux';
 
const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
    useSelector: jest.fn(),
    useDispatch: () => mockDispatch
}))

describe('Post component', () => {
    let root: any;
    let listP: Post[];
    let currentDate = new Date();

    function doBasicChecks() {
        let domElement: ReactWrapper = root.update().find(PostDetail).find('div.PostDetailContainer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostDetailContainer > div.AuthorContainer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();
        domElement = root.find('div.PostDetailContainer > div.InnerContainer');
        expect(domElement.length).toEqual(1);
        expect(domElement.get(0)).toBeTruthy();

    }

    function htmlDecode(input: string){
        var e = document.createElement('textarea');
        e.innerHTML = input;
        return e.childNodes.length === 0 ? "" : e.childNodes[0].nodeValue;
      }

    function createPosts(): Post[] {
        return [{
            author: 'author1',
            createdTime: moment(currentDate).add(-4, 'days').add(-6, 'hours').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 1230,
            id: 'p1',
            postType: PostType.IMAGE,
            thumbnailUrl: '/p1-thumb.jpg',
            title: 'My new post 1',
            subreddit: 'Sub',
            postUrl: '/p1.jpg',
            postHtml: null
        },
        {
            author: 'author1',
            createdTime: moment(currentDate).add(-10, 'days').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 26,
            id: 'p2',
            postType: PostType.VIDEO,
            thumbnailUrl: '/p2-thumb.jpg',
            title: 'My new post 2',
            subreddit: 'Sub2',
            postUrl: '/p2.jpg',
            postHtml: null,
            videoData: {
                url: 'http://myurl.com/video.mpg',
                width: 200,
                height: 200
            }
        },
        {
            author: 'author2',
            createdTime: moment(currentDate).add(-2, 'minutes').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 900,
            id: 'p3',
            postType: PostType.LINK,
            thumbnailUrl: undefined,
            title: 'My new post 3',
            subreddit: 'Sub3',
            postUrl: 'http://www.google.com',
            postHtml: null
        },
        {
            author: 'author3',
            createdTime: moment(currentDate).add(-12, 'seconds').add(+new Date().getTimezoneOffset(), 'minutes').toDate(),
            numberOfComments: 0,
            id: 'p4',
            postType: PostType.SELF,
            thumbnailUrl: undefined,
            title: 'My new post 4',
            subreddit: 'Sub4',
            postUrl: undefined,
            postHtml: `&lt;iframe class="embedly-embed" src="https://cdn.embedly.com/widgets/media.html?src=https%3A%2F%2Fgfycat.com%2Fifr%2Fanotherreddachshund&amp;display_name=Gfycat&amp;url=https%3A%2F%2Fgfycat.com%2Fanotherreddachshund&amp;image=https%3A%2F%2Fthumbs.gfycat.com%2FAnotherRedDachshund-size_restricted.gif&amp;key=ed8fa8699ce04833838e66ce79ba05f1&amp;type=text%2Fhtml&amp;schema=gfycat" width="460" height="574" scrolling="no" title="Gfycat embed" frameborder="0" allow="autoplay; fullscreen" allowfullscreen="true"&gt;&lt;/iframe&gt;`
        }];
    }

    beforeEach((dn) => {
        
        listP = createPosts();

        dn();
    });
 
    test('Show image post detail', () => 
    {
        const p = listP[0];
        (useSelector as any).mockImplementation(() => p);
        
        root = mount(<PostDetail />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: createStore(rootReducer)
            }
        });
        doBasicChecks();
        let domElement: ReactWrapper = root.find('div.PostDetailContainer > div.AuthorContainer');
        expect(domElement.text()).toEqual(p.author);
        domElement = root.find('div.InnerContainer > span.Title');
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.text()).toEqual(p.title);

        domElement = root.find('div.Body > div.Body-content > div.PostPicture');
        expect(domElement.get(0)).toBeTruthy();
        expect((domElement.getDOMNode() as any).style).toHaveProperty('background-image', `url(${p.postUrl})`);

        
    });

    test('Show video post detail', () => 
    {
        const p = listP[1];
        (useSelector as any).mockImplementation(() => p);
        
        root = mount(<PostDetail />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: createStore(rootReducer)
            }
        });
        doBasicChecks();
        let domElement: ReactWrapper = root.find('div.PostDetailContainer > div.AuthorContainer');
        expect(domElement.text()).toEqual(p.author);
        domElement = root.find('div.InnerContainer > span.Title');
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.text()).toEqual(p.title);

        domElement = root.find('div.Body > div.Body-content > video');
        expect(domElement.get(0)).toBeTruthy();
        expect((domElement.getDOMNode() as any)).toHaveProperty('src', p.videoData.url);
    });

    test('Show link post detail', () => 
    {
        const p = listP[2];
        (useSelector as any).mockImplementation(() => p);
        
        root = mount(<PostDetail />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: createStore(rootReducer)
            }
        });
        doBasicChecks();

        const domElement = root.find('div.Body > div.Body-content > a.Url');
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.text()).toEqual(p.postUrl);
    });

    test('Show embbed html post detail', () => 
    {
        const p = listP[3];
        (useSelector as any).mockImplementation(() => p);
        
        root = mount(<PostDetail />, {
            wrappingComponent: Provider,
            wrappingComponentProps: {
                store: createStore(rootReducer)
            }
        });
        doBasicChecks();

        const domElement = root.find('div.Body > div.Body-content.HtmlBody');
        expect(domElement.get(0)).toBeTruthy();
        expect(domElement.getDOMNode().innerHTML.replace(/&amp;/g, '&')).toEqual(htmlDecode(p.postHtml));
    });

});