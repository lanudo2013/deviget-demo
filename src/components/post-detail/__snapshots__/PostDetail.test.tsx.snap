// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component Show embbed html post detail 1`] = `
<div
  className="PostDetailContainer"
>
  <div
    className="AuthorContainer"
  >
    <span
      className="Author"
    >
      author3
    </span>
  </div>
  <div
    className="InnerContainer"
    style={Object {}}
  >
    <span
      className="Title"
      style={Object {}}
    >
      My new post 4
    </span>
    <div
      className="Body"
    >
      <div
        className="Body-content NotCentered HtmlBody"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<iframe class=\\"embedly-embed\\" src=\\"https://cdn.embedly.com/widgets/media.html?src=https%3A%2F%2Fgfycat.com%2Fifr%2Fanotherreddachshund&display_name=Gfycat&url=https%3A%2F%2Fgfycat.com%2Fanotherreddachshund&image=https%3A%2F%2Fthumbs.gfycat.com%2FAnotherRedDachshund-size_restricted.gif&key=ed8fa8699ce04833838e66ce79ba05f1&type=text%2Fhtml&schema=gfycat\\" width=\\"460\\" height=\\"574\\" scrolling=\\"no\\" title=\\"Gfycat embed\\" frameborder=\\"0\\" allow=\\"autoplay; fullscreen\\" allowfullscreen=\\"true\\"></iframe>",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Post component Show image post detail 1`] = `
<div
  className="PostDetailContainer"
>
  <div
    className="AuthorContainer"
  >
    <span
      className="Author"
    >
      author1
    </span>
  </div>
  <div
    className="InnerContainer"
    style={
      Object {
        "maxHeight": "calc(100% - 6em)",
      }
    }
  >
    <span
      className="Title"
      style={Object {}}
    >
      My new post 1
    </span>
    <div
      className="Body"
    >
      <div
        className="Body-content"
      >
        <div
          className="PostPicture"
          style={
            Object {
              "backgroundImage": "url(/p1.jpg)",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Post component Show link post detail 1`] = `
<div
  className="PostDetailContainer"
>
  <div
    className="AuthorContainer"
  >
    <span
      className="Author"
    >
      author2
    </span>
  </div>
  <div
    className="InnerContainer"
    style={
      Object {
        "maxHeight": "calc(100% - 6em)",
      }
    }
  >
    <span
      className="Title"
      style={Object {}}
    >
      My new post 3
    </span>
    <div
      className="Body"
    >
      <div
        className="Body-content NotCentered"
      >
        <a
          className="Url"
          href="http://www.google.com"
          target="_blank"
        >
          http://www.google.com
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Post component Show video post detail 1`] = `
<div
  className="PostDetailContainer"
>
  <div
    className="AuthorContainer"
  >
    <span
      className="Author"
    >
      author1
    </span>
  </div>
  <div
    className="InnerContainer"
    style={
      Object {
        "maxHeight": "calc(100% - 6em)",
      }
    }
  >
    <span
      className="Title"
      style={Object {}}
    >
      My new post 2
    </span>
    <div
      className="Body"
    >
      <div
        className="Body-content"
      >
        <video
          autoPlay={true}
          className="Video"
          controls={true}
          src="http://myurl.com/video.mpg"
        />
      </div>
    </div>
  </div>
</div>
`;
